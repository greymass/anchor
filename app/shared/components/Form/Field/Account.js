// @flow
import React, { Component } from 'react';
import { I18n } from 'react-i18next';
import { Form, Input } from 'semantic-ui-react';

import debounce from 'lodash/debounce';

export default class FormFieldAccount extends Component<Props> {
  state = { value: '' };
  onChange = debounce((e, { name, value }) => {
    const parsed = value;
    this.setState({
      value: parsed
    }, () => {
      this.props.onChange(e, { name, value: parsed });
    });
  }, 300)
  render() {
    const {
      autoFocus,
      disabled,
      icon,
      label,
      loading,
      name,
      value
    } = this.props;
    return (
      <Form.Field
        autoFocus={autoFocus}
        control={Input}
        disabled={disabled}
        fluid
        icon={icon}
        label={label}
        loading={loading}
        name={name}
        onChange={this.onChange}
        defaultValue={value}
      />
    );
  }
}
