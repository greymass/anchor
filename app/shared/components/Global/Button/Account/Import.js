// @flow
import React, { Component } from 'react';
import { translate } from 'react-i18next';
import { Button } from 'semantic-ui-react';

import GlobalModalAccountImport from '../../Modal/Account/Import';

class GlobalButtonAccountImport extends Component<Props> {
  state = {
    open: true
  }

  onOpen = () => this.setState({ open: true });
  onClose = () => this.setState({ open: false });

  render() {
    const {
      t
    } = this.props;
    const {
      open
    } = this.state;
    return (
      <GlobalModalAccountImport
        onClose={this.onClose}
        open={open}
        trigger={(
          <Button
            color="purple"
            content={t('wallet_panel_wallet_unlock')}
            fluid
            icon="unlock"
            onClick={this.onOpen}
          />
        )}
      />
    );
  }
}

export default translate('global')(GlobalButtonAccountImport);
