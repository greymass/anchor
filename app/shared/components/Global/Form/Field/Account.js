// @flow
import React, { Component } from 'react';
import { Form, Input } from 'semantic-ui-react';
import { withTranslation } from 'react-i18next';
import { sortBy } from 'lodash';
import { get } from 'dot-prop-immutable';

export class GlobalFormFieldAccount extends Component<Props> {
  constructor(props) {
    super(props);
    this.state = {
      fieldOption: 'manual',
      maxlength: props.maxlength || 12,
      value: props.value
    };
  }
  onChange = (e, { name, value }) => {
    const parsed = value.trim().toLowerCase();
    let valid = false;
    switch (this.props.rules) {
      case 'generic':
        valid = !!(parsed.match(new RegExp("^[a-z12345]{" + this.state.maxlength + "}$", "g")));
        break;
      default:
        valid = !!(parsed.match(new RegExp("^[a-z12345.]{1," + this.state.maxlength + "}$", "g")));
        break;
    }
    this.setState({
      value: parsed
    }, () => {
      this.props.onChange(e, {
        name,
        value: parsed,
        valid
      });
    });
  };
  reset = () => this.setState({ value: '' });

  handleRadioChange = (e, { value }) => {
    const {
      name
    } = this.props;

    this.setState({
      fieldOption: value
    }, () => {
      this.onChange(e, { name, value: '' });
    });
  };

  render() {
    const {
      app,
      autoFocus,
      contacts,
      disabled,
      enableContacts,
      enableExchanges,
      showErrorOnInput = false,
      fluid,
      icon,
      label,
      labelPosition,
      loading,
      name,
      t,
      width
    } = this.props;

    const { value } = this.state;

    let { fieldOption } = this.state;

    const exchangeAccounts = get(app, 'constants.exchanges') || [];

    if (contacts && contacts.includes(value)) {
      fieldOption = 'contacts';
    } else if (exchangeAccounts && exchangeAccounts.includes(value)) {
      fieldOption = 'exchanges';
    }

    let dropdownOptions;

    if (fieldOption === 'contacts') {
      dropdownOptions = sortBy(contacts, c => c.accountName).map((contact) => ({
        value: contact.accountName,
        text: `${contact.accountName} ${contact.label ? (`(${contact.label})`) : ''}`
      }));
    } else if (fieldOption === 'exchanges') {
      dropdownOptions = sortBy(exchangeAccounts).map((exchangeAccount) => ({
        value: exchangesInfo[exchangeAccount].account,
        text: exchangesInfo[exchangeAccount].name,
      }));
    }

    const availableOptions = ['manual'];
    if (enableExchanges) {
      availableOptions.push('exchanges');
    }
    if (enableContacts && contacts && contacts.length > 0) {
      availableOptions.push('contacts');
    }

    const showOptions = (availableOptions.length > 1);

    const inlineLabel = (!showOptions ? label : false);

    const inputField = (
      <Form.Field
        autoFocus={autoFocus}
        control={Input}
        disabled={disabled}
        fluid={fluid}
        icon={icon}
        label={inlineLabel}
        labelPosition={labelPosition}
        loading={loading}
        name={name}
        onChange={this.onChange}
        ref={ref => { this.input = ref; }}
        value={value}
        width={width}
        error={showErrorOnInput}
      />
    );

    if (!showOptions) {
      return inputField;
    }

    return (
      <Form.Field>
        {(showOptions)
          ? (
            <Form.Group inline>
              <label>{t('global_form_field_account_options')}</label>
              {availableOptions.map((option) => (
                <Form.Radio
                  label={t(`global_form_field_account_${option}`)}
                  key={option}
                  name="inputRadioOptions"
                  style={{ marginLeft: '10px' }}
                  value={option}
                  checked={fieldOption === option}
                  onChange={this.handleRadioChange}
                />
              ))}
            </Form.Group>
          ) : <br />}
        <label htmlFor={name}>
          <strong>{label}</strong>
          {(fieldOption === 'manual')
          ? (
            inputField
          ) : ''}

          {(fieldOption !== 'manual')
          ? (
            <Form.Dropdown
              value={value}
              fluid
              name={name}
              onChange={this.onChange}
              options={dropdownOptions}
              placeholder={
                fieldOption === 'exchanges' ?
                  t('global_form_placeholder_exchange') :
                  t('global_form_placeholder_contact')
              }
              selection
            />
          ) : ''}
        </label>
      </Form.Field>
    );
  }
}

export default withTranslation('global')(GlobalFormFieldAccount);

const exchangesInfo = {
  binancecleos: {
    account: 'binancecleos',
    name: 'Binance',
  },
  bitfinexdep1: {
    account: 'bitfinexdep1',
    name: 'Bitfinex',
  },
  krakenkraken: {
    account: 'krakenkraken',
    name: 'Kraken',
  },
};
