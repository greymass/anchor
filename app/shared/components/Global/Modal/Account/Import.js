// @flow
import React, { Component } from 'react';
import { translate } from 'react-i18next';
import { Button, Header, Icon, Modal, Tab } from 'semantic-ui-react';

import GlobalModalAccountImportCold from '../../../../containers/Global/Account/Import/Cold';
import GlobalModalAccountImportHot from '../../../../containers/Global/Account/Import/Hot';
import GlobalModalAccountImportWatch from '../../../../containers/Global/Account/Import/Watch';

class GlobalModalAccountImport extends Component<Props> {
  getPanes = () => {
    const {
      onClose,
      settings,
      t
    } = this.props;
    const panes = [];

    const hotWallet = {
      menuItem: t('global_modal_account_import_hot_wallet'),
      render: () => <GlobalModalAccountImportHot onClose={onClose} />
    };
    const watchWallet = {
      menuItem: t('global_modal_account_import_watch_wallet'),
      render: () => <GlobalModalAccountImportWatch onClose={onClose} />
    };
    const coldWallet = {
      menuItem: t('global_modal_account_import_cold_wallet'),
      render: () => <GlobalModalAccountImportCold onClose={onClose} />
    };

    switch (settings.walletMode) {
      case 'cold': {
        panes.push(coldWallet);
        break;
      }
      case 'watch': {
        panes.push(watchWallet, hotWallet);
        break;
      }
      default: {
        panes.push(hotWallet, watchWallet);
      }
    }

    return panes;
  }

  render() {
    const {
      onClose,
      open,
      t,
      trigger
    } = this.props;
    const panes = this.getPanes();
    return (
      <Modal
        centered={false}
        trigger={trigger}
        onClose={onClose}
        open={open}
        size="small"
      >
        <Header
          content={t('global_button_account_import_action')}
          icon="users"
          subheader={t('global_button_account_import_action')}
        />
        <Tab
          menu={{
            attached: true,
            inverted: true,
            fluid: true,
            pointing: true,
            style: {
              margin: 0
            }
          }}
          panes={panes}
          defaultActiveIndex={0}
        />
      </Modal>
    );
  }
}

export default translate('global')(GlobalModalAccountImport);
