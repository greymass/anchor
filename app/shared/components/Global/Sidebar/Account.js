// @flow
import React, { Component } from 'react';
import { translate } from 'react-i18next';

import { Button, Divider, Dropdown, Header, Input, Segment } from 'semantic-ui-react';

import GlobalButtonAccountImport from '../Button/Account/Import';

class GlobalSidebarAccount extends Component<Props> {
  render() {
    const {
      wallet
    } = this.props;
    const tagOptions = [
      {
        text: 'accountname',
        value: 'accountname',
        label: { color: 'black', empty: true, circular: true },
      },
    ];
    return (
      <Button.Group
        basic
        fluid
      >
        <Dropdown
          button
          className="icon"
          icon="users"
          labeled
          text={wallet.account}
        >
          <Dropdown.Menu>
            <Input
              icon="search"
              iconPosition="left"
              className="search"
            />
            <Dropdown.Menu scrolling>
              {tagOptions.map(option => <Dropdown.Item key={option.value} {...option} />)}
            </Dropdown.Menu>
          </Dropdown.Menu>
        </Dropdown>
        <GlobalButtonAccountImport
        />
      </Button.Group>
    );
  }
}

export default translate('sidebar')(GlobalSidebarAccount);
