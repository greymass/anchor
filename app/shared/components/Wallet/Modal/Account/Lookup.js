// @flow
import React, { Component } from 'react';
import { I18n } from 'react-i18next';
import { Button, Form, Header, Icon, Input, List, Message, Modal, Segment } from 'semantic-ui-react';

const ecc = require('eosjs-ecc');

export default class WalletModalAccountLookup extends Component<Props> {
  state = {
    key: '',
    open: false,
    valid: false
  }

  onChange = (e, { value }) => {
    this.setState({
      key: value,
      valid: (ecc.isValidPublic(value) === true)
    }, () => {
      const { key, valid } = this.state;
      if (valid) {
        this.props.actions.getAccountByKey(key);
      } else {
        this.props.actions.clearAccountByKey();
      }
    });
  }

  onKeyPress = (e) => {
    if (e.key === 'Enter') {
      this.onSubmit();
    }
  }

  onOpen = () => this.setState({ open: true });
  onClose = () => this.setState({ open: false });

  component

  render() {
    const {
      accounts,
      validate
    } = this.props;
    const {
      key,
      open,
      valid
    } = this.state;
    const {
      __lookups
    } = accounts;
    return (
      <I18n ns="wallet">
        {
          (t) => (
            <Modal
              trigger={(
                <a
                  onClick={this.onOpen}
                  role="button"
                >
                  {t('wallet_account_lookup_modal_trigger')}
                </a>
              )}
              onClose={this.onClose}
              open={open}
              size="tiny"
            >
              <Header icon="unlock" content={t('wallet_account_lookup_modal_title')} />
              <Modal.Content>
                <h3>{t('wallet_account_lookup_modal_description')}</h3>
                <Form.Field
                  autoFocus
                  control={Input}
                  fluid
                  label={t('wallet_account_lookup_modal_field_label')}
                  onChange={this.onChange}
                  onKeyPress={this.onKeyPress}
                />
                {(valid)
                  ? (
                    <Segment padded>
                      <Header size="small">
                        {t('wallet_account_lookup_modal_accounts_found')}
                      </Header>
                      <List divided relaxed>
                        {(!__lookups.length)
                          ? <List.Item key="none found" content={t('wallet_account_lookup_modal_accounts_found_none')} />
                          : null
                        }
                        {__lookups.map((account) => <List.Item key={account} content={account} />)}
                      </List>
                    </Segment>
                  )
                  : (
                    <span>
                      {(key)
                        ? (
                          <Segment padded>
                            <Header>
                              {t('wallet_account_lookup_modal_key_invalid')}
                            </Header>
                          </Segment>
                        )
                        : null
                      }
                    </span>
                  )
                }
                {(validate.WALLET_PASSWORD === 'FAILURE')
                  ? (
                    <Segment padded>
                      <Message
                        content={t('wallet_account_lookup_failure_content')}
                        error
                        header={t('wallet_account_lookup_failure_header')}
                        icon="warning circle"
                      />
                    </Segment>
                  ) : null
                }
              </Modal.Content>
              <Modal.Actions>
                <Button
                  onClick={this.onClose}
                >
                  <Icon name="x" /> {t('close')}
                </Button>
              </Modal.Actions>
            </Modal>
          )
        }
      </I18n>
    );
  }
}
