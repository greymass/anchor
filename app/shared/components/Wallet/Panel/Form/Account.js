// @flow
import React, { Component } from 'react';
import { I18n } from 'react-i18next';
import { Form, Input } from 'semantic-ui-react';
import WalletModalAccountLookup from '../../Modal/Account/Lookup';

export default class WalletPanelFormAccount extends Component<Props> {
  render() {
    const {
      actions,
      accounts,
      onChange,
      validate,
      value
    } = this.props;
    return (
      <I18n ns="wallet">
        {
          (t) => (
            <Form.Field
              control={Input}
              fluid
              icon={(validate.NODE === 'SUCCESS' && validate.ACCOUNT === 'SUCCESS') ? 'checkmark' : 'x'}
              label={(
                <span>
                  <span style={{float: 'right'}}>
                    <WalletModalAccountLookup
                      actions={actions}
                      accounts={accounts}
                      validate={validate}
                    />
                  </span>
                  {t('wallet_panel_form_account')}
                </span>
              )}
              loading={(validate.ACCOUNT === 'PENDING')}
              name="account"
              onChange={onChange}
              defaultValue={value}
            />
          )
        }
      </I18n>
    );
  }
}
