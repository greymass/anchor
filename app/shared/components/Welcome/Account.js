// @flow
import React, { Component } from 'react';
import { Header, Segment } from 'semantic-ui-react';
import { translate } from 'react-i18next';
import WelcomeAccountContainer from '../../containers/Welcome/Account';

class WelcomeAccount extends Component<Props> {
  render() {
    const {
      onStageSelect,
      t
    } = this.props;
    return (
      <Segment
        size="small"
        stacked
      >
        <Header>
          {t('welcome_stage')} #2: {t('welcome_stage_account')}
        </Header>
        <WelcomeAccountContainer
          onStageSelect={onStageSelect}
        />
      </Segment>
    );
  }
}

export default translate('welcome')(WelcomeAccount);
