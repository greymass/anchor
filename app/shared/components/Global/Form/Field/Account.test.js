import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { GlobalFormFieldAccount } from './Account';

const mockProps = {
  bytes: 1000,
  app: {},
  contacts: [{ accountName: 'teamgreymass' }],
  enableContacts: true,
  enableExchanges: true,
  showErrorOnInput: true,
  icon: 'warning',
  label: 'testing account field',
  name: 'account_field',
  t: () => null,
  width: 10,
};

describe('GlobalFormFieldAccount', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<GlobalFormFieldAccount {...mockProps} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
