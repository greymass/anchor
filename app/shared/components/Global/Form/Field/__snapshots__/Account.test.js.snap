// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalFormFieldAccount renders correctly 1`] = `
<FormField>
  <FormGroup
    inline={true}
  >
    <label />
    <FormRadio
      as={[Function]}
      checked={true}
      control={[Function]}
      key="manual"
      label={null}
      name="inputRadioOptions"
      onChange={[Function]}
      style={
        Object {
          "marginLeft": "10px",
        }
      }
      value="manual"
    />
    <FormRadio
      as={[Function]}
      checked={false}
      control={[Function]}
      key="exchanges"
      label={null}
      name="inputRadioOptions"
      onChange={[Function]}
      style={
        Object {
          "marginLeft": "10px",
        }
      }
      value="exchanges"
    />
    <FormRadio
      as={[Function]}
      checked={false}
      control={[Function]}
      key="contacts"
      label={null}
      name="inputRadioOptions"
      onChange={[Function]}
      style={
        Object {
          "marginLeft": "10px",
        }
      }
      value="contacts"
    />
  </FormGroup>
  <label
    htmlFor="account_field"
  >
    <strong>
      testing account field
    </strong>
    <FormField
      control={[Function]}
      error={true}
      icon="warning"
      label={false}
      name="account_field"
      onChange={[Function]}
      width={10}
    />
  </label>
</FormField>
`;
